### PR TITLE
[TradingModes] init optimize_initial_portfolio

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot-Trading/pull/984
todo:
- ✅ DCA
- grid

### CODE REVIEW

File: Trading/Exchange/kucoin/kucoin_exchange.py [hunk 1]
```diff
@@ -161,8 +161,11 @@ async def get_account_id(self, **kwargs: dict) -> str:
         except ccxt.AuthenticationError:
             # when api key is wrong
             raise
-        except ccxt.ExchangeError:
+        except ccxt.ExchangeError as err:
             # ExchangeError('kucoin This user is not a master user')
+            if "not a master user" not in str(err):
```
[the remark this answers is not in the document]
Comment (GuillaumeDSM) [reply Trading/Exchange/kucoin/kucoin_exchange.py:166]: it's a check to be sure kucoin doesn't change its behavior ^^

File: Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py [hunk 2]
```diff
@@ -251,6 +253,92 @@ def get_is_symbol_wildcard(cls) -> bool:
...
+                for symbol in self.exchange_manager.exchange_config.traded_symbol_pairs:
+                    if producer.get_symbol_trading_config(symbol) is not None:
+                        configured_pairs.append(symbol)
+                        pair_bases.add(symbol_util.parse_symbol(symbol).base)
+                        for order in self.exchange_manager.exchange_personal_data.orders_manager.get_open_orders(
+                            symbol=symbol
+                        ):
+                            if not (order.is_cancelled() or order.is_closed()):
+                                await self.cancel_order(order)
+                                cancelled_orders.append(order)
```
[the remark this answers is not in the document]
Comment (GuillaumeDSM) [reply Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py:293]: there is but we need to do it with other conditions in if that are irrelevant here and using `self.cancel_order` so I thing it's better to do it this way

File: Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py [hunk 3]
```diff
@@ -251,6 +253,92 @@ def get_is_symbol_wildcard(cls) -> bool:
     def set_default_config(self):
         raise RuntimeError(f"Impossible to start {self.get_name()} without a valid configuration file.")
 
+    async def optimize_initial_portfolio(self, sellable_assets: list, tickers: dict = None) -> (list, dict):
```
[the remark this answers is not in the document]
Comment (GuillaumeDSM) [reply Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py:256]: indeed ! 👍

File: Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py [hunk 4]
```diff
@@ -271,73 +271,108 @@ async def optimize_initial_portfolio(self, sellable_assets: list, tickers: dict
...
-                    )
+                part_2_orders = await self._buy_assets(
+                    producer, pair_bases, common_quote, converted_quote_amount_per_symbol, tickers
+                )
+
                 await trading_modes.notify_portfolio_optimization_complete()
-                return [cancelled_orders, part_1_orders, part_2_orders], tickers
+                return [cancelled_orders, part_1_orders, part_2_orders]
+
+    async def _cancel_associated_orders(self, producer, pair_bases) -> list:
```
Comment (Herklos) [inline Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py:309]: :+1:

File: Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py [hunk 5]
```diff
@@ -251,6 +253,110 @@ def get_is_symbol_wildcard(cls) -> bool:
     def set_default_config(self):
         raise RuntimeError(f"Impossible to start {self.get_name()} without a valid configuration file.")
 
+    async def single_exchange_process_optimize_initial_portfolio(
```
Comment (Herklos) [inline Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py:256]: :+1:

File: Trading/Mode/staggered_orders_trading_mode/tests/test_staggered_orders_trading_mode.py [hunk 6]
```diff
@@ -1339,6 +1346,53 @@ async def test_ensure_current_price_in_limit_parameters():
...
+        part_2_order = part_2_orders[0]
+        assert isinstance(part_2_order, trading_personal_data.BuyMarketOrder)
+        assert part_2_order.created_last_price == decimal.Decimal("1000")
+        assert part_2_order.origin_quantity == decimal.Decimal("5.545")    # 50% of funds
+        assert part_2_order.status == trading_enums.OrderStatus.FILLED
+
+        # check portfolio is rebalanced
+        final_portfolio = exchange_manager.exchange_personal_data.portfolio_manager.portfolio.portfolio
+        assert final_portfolio["BTC"].available == decimal.Decimal('5.539455')  # 5.545 - fees
+        assert final_portfolio["USD"].available == decimal.Decimal("5545")
```
Comment (Herklos) [inline Trading/Mode/staggered_orders_trading_mode/tests/test_staggered_orders_trading_mode.py:1393]: :+1:

**Review by Herklos — APPROVED** [review]

**Pull request merged**